### PR TITLE
Added a center_offset property to both plane primitive and quad primitive

### PR DESCRIPTION
**Pull request by QbieShay**
Adds offset to plane and quad meshes, useful for VFX.

*Bugsquad edit:* 3.x version: #48763

### CODE REVIEW

**Comment by Calinou** [comment]
Out of curiosity, why not move the plane/quad mesh itself? Or is this for use in particles?

**Comment by QbieShay** [comment]
@Calinou particles exactly :)
simplest example that comes to mind is if you want to have some rays coming out in all directions. Right now you need to make your own custom mesh with a pivot at the bottom of the quad, but with this addition you don't need a custom mesh.

**Comment by Calinou** [comment]
I wonder if this property should be added to all primitive meshes, or just QuadMesh or PlaneMesh. There's a PrimitiveMesh base class after all.

**Comment by QbieShay** [comment]
I have thought about it, but i wasn't sure of the usecase for other meshes. It shouldn't be a problem to add it later, if the name is kept, right?

**Comment by Calinou** [comment]
> I have thought about it, but i wasn't sure of the usecase for other meshes. It shouldn't be a problem to add it later, if the name is kept, right?

I think it's OK to make a property more general in future `3.x` releases, but I'm not sure.

**Review by Calinou — APPROVED** [review]
Feature looks good to me.

File: scene/resources/primitive_meshes.h [hunk 1]
```diff
@@ -211,6 +211,7 @@ class PlaneMesh : public PrimitiveMesh {
 	Size2 size = Size2(2.0, 2.0);
 	int subdivide_w = 0;
 	int subdivide_d = 0;
+	Vector3 center_offset = Vector3(0.0, 0.0, 0.0);
```
Comment (akien-mga) [inline scene/resources/primitive_meshes.h:214]: `0, 0, 0` is the default value for `Vector3`, so you don't need to assign anything.

File: scene/resources/primitive_meshes.cpp [hunk 2]
```diff
@@ -1026,6 +1029,14 @@ int PlaneMesh::get_subdivide_depth() const {
...
 }
 
+void PlaneMesh::set_center_offset(const Vector3 p_offset) {
+	center_offset = p_offset;
+	_request_update();
+}
+
+Vector3 PlaneMesh::get_center_offset() const {
+	return center_offset;
+}
```
Comment (akien-mga) [inline scene/resources/primitive_meshes.cpp:1039]: Keep a separation line before the constructor.

File: scene/resources/primitive_meshes.cpp [hunk 3]
```diff
@@ -258,7 +258,7 @@ PrimitiveMesh::~PrimitiveMesh() {
 }
 
 /**
-	CapsuleMesh
+    CapsuleMesh
```
Comment (akien-mga) [inline scene/resources/primitive_meshes.cpp:261]: If changing this, would be good to make it match others (2 spaces).

File: scene/resources/primitive_meshes.h [hunk 4]
```diff
@@ -274,6 +278,7 @@ class QuadMesh : public PrimitiveMesh {
 
 private:
 	Size2 size = Size2(1.0, 1.0);
+	Vector3 center_offset = Vector3(0.0, 0.0, 0.0);
```
Comment (akien-mga) [inline scene/resources/primitive_meshes.h:281]: Same here.

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**